### PR TITLE
Add support for directly iterating over arrays

### PR DESCRIPTION
**Pull request by arteam**
Avoid creating a stream and a spliterator when a simple iteration over an array with a counter would be sufficient.

### CODE REVIEW

**Review by grcevski — COMMENTED** [review]
This looks really good to me! I think we should add a bit more tests:

- Write a test for a contract of what's the expected behaviour for passing a 'null' array to the iterator. Perhaps we should call Objects.requireNonNull with a custom message saying we are not allowing null arrays wrapped in an iterator? This will also retain the existing behaviour of the code that currently creates a spliterator() with Arrays.stream.
- The Iterator interface comes with a default method forEachRemaining(). We should probably have a test that ensures this method works with our custom iterator.

Thanks!

**Comment by arteam** [comment]
@elasticmachine update branch

File: server/src/test/java/org/elasticsearch/common/collect/IteratorsTests.java [hunk 1]
```diff
@@ -110,6 +111,37 @@ public void testNullIterator() {
...
+        Iterator<Integer> iterator = Iterators.forArray(array);
+        for (int i = 0; i < array.length; i++) {
+            iterator.next();
+        }
+
+        expectThrows(NoSuchElementException.class, iterator::next);
+    }
+
+    private static Integer[] randomIntegerArray() {
+        return Randomness.get().ints(randomIntBetween(1, 1000)).boxed().toArray(Integer[]::new);
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/common/collect/IteratorsTests.java:142]: I think it's important also to verify that it behaves correctly with empty arrays:

```suggestion
        return Randomness.get().ints(randomIntBetween(0, 1000)).boxed().toArray(Integer[]::new);
```

File: server/src/test/java/org/elasticsearch/common/collect/IteratorsTests.java [hunk 2]
```diff
@@ -110,6 +111,37 @@ public void testNullIterator() {
...
     }
 
+    public void testArrayIterator() {
+        Integer[] array = randomIntegerArray();
+        Iterator<Integer> iterator = Iterators.forArray(array);
+
+        int i = 0;
+        while (iterator.hasNext()) {
+            assertEquals(array[i++], iterator.next());
+        }
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/common/collect/IteratorsTests.java:121]: I think we should also assert that the array has no more elements here.

**Review by DaveCTurner — COMMENTED** [review]
I left a couple of small comments about the unit tests.

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine update branch

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by arteam** [comment]
Thanks, David!

**Pull request merged**